File: src/rdp_file_parser.cpp
/**
 * RDP File Parser Implementation
 * 
 * Parses Microsoft Remote Desktop Connection (.rdp/.rdpw) files.
 */

#include "rdp_file_parser.hpp"
#include "json_utils.hpp"
#include "logger.hpp"

#include <fstream>
#include <sstream>
#include <algorithm>
#include <iostream>
#include <cctype>
#include <functional>
#include <unordered_map>
#include <format>

// ============================================================================
// Table-driven parser: maps lowercase key → setter on RDPFileData
// ============================================================================

namespace {

using Setter = std::function<void(RDPFileData&, const std::string&)>;

// Helper: parse string with unescaping
Setter str_setter(std::string RDPFileData::* field) {
    return [field](RDPFileData& d, const std::string& v) { d.*field = v; };
}

// Helper: parse "0"/"1" into bool
Setter bool_setter(bool RDPFileData::* field) {
    return [field](RDPFileData& d, const std::string& v) { d.*field = (v == "1"); };
}

// Helper: parse integer with safe fallback
Setter int_setter(int RDPFileData::* field) {
    return [field](RDPFileData& d, const std::string& v) {
        try { d.*field = std::stoi(v); }
        catch (...) { /* keep default */ }
    };
}

const std::unordered_map<std::string, Setter>& rdp_key_map() {
    static const std::unordered_map<std::string, Setter> map = {
        // Basic connection
        {"full address",              str_setter(&RDPFileData::full_address)},
        {"alternate full address",    str_setter(&RDPFileData::alternate_full_address)},
        {"server port",               int_setter(&RDPFileData::server_port)},
        {"username",                  str_setter(&RDPFileData::username)},
        {"domain",                    str_setter(&RDPFileData::domain)},
        // Authentication
        {"prompt for credentials",    bool_setter(&RDPFileData::prompt_for_credentials)},
        {"promptcredentialonce",      bool_setter(&RDPFileData::prompt_credential_once)},
        {"authentication level",      int_setter(&RDPFileData::authentication_level)},
        // Azure AD
        {"targetisaadjoined",         bool_setter(&RDPFileData::target_is_aad_joined)},
        {"enablerdsaadauth",          bool_setter(&RDPFileData::enable_rds_aad_auth)},
        {"aadtenantid",               str_setter(&RDPFileData::aad_tenant_id)},
        // Gateway
        {"gatewayhostname",           str_setter(&RDPFileData::gateway_hostname)},
        {"gatewayusagemethod",        int_setter(&RDPFileData::gateway_usage_method)},
        {"gatewayprofileusagemethod",  int_setter(&RDPFileData::gateway_profile_usage_method)},
        {"gatewaycredentialssource",   int_setter(&RDPFileData::gateway_credentials_source)},
        {"gatewaybrokeringtype",       int_setter(&RDPFileData::gateway_brokering_type)},
        // AVD / WVD
        {"loadbalanceinfo",           str_setter(&RDPFileData::load_balance_info)},
        {"wvd endpoint pool",         str_setter(&RDPFileData::wvd_endpoint_pool)},
        {"armpath",                   str_setter(&RDPFileData::arm_path)},
        {"workspace id",             str_setter(&RDPFileData::workspace_id)},
        {"resourceprovider",          str_setter(&RDPFileData::resource_provider)},
        {"geo",                       str_setter(&RDPFileData::geo)},
        {"diagnosticserviceurl",      str_setter(&RDPFileData::diagnostic_service_url)},
        {"hubdiscoverygeourl",        str_setter(&RDPFileData::hub_discovery_geo_url)},
        {"activityhint",              str_setter(&RDPFileData::activity_hint)},
        // Remote app
        {"remoteapplicationprogram",  str_setter(&RDPFileData::remote_application_program)},
        {"remotedesktopname",         str_setter(&RDPFileData::remote_desktop_name)},
        {"remoteapplicationmode",     int_setter(&RDPFileData::remote_application_mode)},
        // Display
        {"desktopwidth",              int_setter(&RDPFileData::desktop_width)},
        {"desktopheight",             int_setter(&RDPFileData::desktop_height)},
        {"screen mode id",           int_setter(&RDPFileData::screen_mode_id)},
        {"smart sizing",             bool_setter(&RDPFileData::smart_sizing)},
        {"dynamic resolution",       bool_setter(&RDPFileData::dynamic_resolution)},
        {"singlemoninwindowedmode",   int_setter(&RDPFileData::single_mon_in_windowed_mode)},
        // Redirection
        {"redirectclipboard",         bool_setter(&RDPFileData::redirect_clipboard)},
        {"redirectprinters",          bool_setter(&RDPFileData::redirect_printers)},
        {"redirectsmartcards",        bool_setter(&RDPFileData::redirect_smart_cards)},
        {"redirectcomports",          bool_setter(&RDPFileData::redirect_com_ports)},
        {"redirectlocation",          bool_setter(&RDPFileData::redirect_location)},
        {"drivestoredirect",          str_setter(&RDPFileData::drives_to_redirect)},
        {"devicestoredirect",         str_setter(&RDPFileData::devices_to_redirect)},
        {"camerastoredirect",         str_setter(&RDPFileData::cameras_to_redirect)},
        {"usbdevicestoredirect",      str_setter(&RDPFileData::usb_devices_to_redirect)},
        // Audio
        {"audiomode",                 int_setter(&RDPFileData::audio_mode)},
        {"audiocapturemode",          int_setter(&RDPFileData::audio_capture_mode)},
        // Security
        {"signscope",                 str_setter(&RDPFileData::sign_scope)},
        {"signature",                 str_setter(&RDPFileData::signature)},
    };
    return map;
}

} // namespace

std::optional<RDPFileData> RDPFileParser::parse_file(std::string_view filepath) {
    std::ifstream file(std::string{filepath});
    if (!file.is_open()) {
        LOG_ERROR("RDPFileParser", "Failed to open file: " << filepath);
        return std::nullopt;
    }
    
    std::stringstream buffer;
    buffer << file.rdbuf();
    file.close();
    
    std::string content = buffer.str();
    auto result = parse_content(content);
    return result;
}

std::optional<RDPFileData> RDPFileParser::parse_content(std::string_view content) {
    RDPFileData data;
    
    std::istringstream stream{std::string{content}};
    std::string line;
    
    while (std::getline(stream, line)) {
        // Remove trailing CR if present (Windows line endings)
        if (!line.empty() && line.back() == '\r') {
            line.pop_back();
        }
        
        // Skip empty lines
        if (line.empty()) continue;
        
        std::string key;
        char type;
        std::string value;
        
        if (!parse_line(line, key, type, value)) {
            continue;  // Skip lines that don't match the format
        }
        
        // Convert key to lowercase for comparison
        std::string key_lower = key;
        std::transform(key_lower.begin(), key_lower.end(), key_lower.begin(), ::tolower);
        
        // Look up and apply the setter from the table
        const auto& map = rdp_key_map();
        auto it = map.find(key_lower);
        if (it != map.end()) {
            it->second(data, unescape_value(value));
        }
    }
    
    // Validate that we have at least a full address or gateway
    if (data.full_address.empty() && data.gateway_hostname.empty()) {
        LOG_ERROR("RDPFileParser", "No valid connection address found");
        return std::nullopt;
    }
    
    LOG_INFO("RDPFileParser", "Parsed RDP file successfully");
    LOG_DEBUG("RDPFileParser", "  Full address: " << data.full_address);
    LOG_DEBUG("RDPFileParser", "  Gateway: " << data.gateway_hostname);
    LOG_DEBUG("RDPFileParser", "  AVD connection: " << (data.is_avd_connection() ? "yes" : "no"));
    LOG_DEBUG("RDPFileParser", "  AAD auth required: " << (data.requires_aad_auth() ? "yes" : "no"));
    
    return data;
}

bool RDPFileParser::parse_line(const std::string& line, 
                               std::string& key, 
                               char& type, 
                               std::string& value) {
    // RDP file format: key:type:value
    // Where type is 's' for string, 'i' for integer, 'b' for binary
    
    size_t first_colon = line.find(':');
    if (first_colon == std::string::npos) {
        return false;
    }
    
    size_t second_colon = line.find(':', first_colon + 1);
    if (second_colon == std::string::npos) {
        return false;
    }
    
    key = line.substr(0, first_colon);
    
    std::string type_str = line.substr(first_colon + 1, second_colon - first_colon - 1);
    if (type_str.empty()) {
        return false;
    }
    type = type_str[0];
    
    value = line.substr(second_colon + 1);
    
    return true;
}

std::string RDPFileParser::unescape_value(const std::string& value) {
    // RDP files may contain URL-encoded or escaped characters
    // For now, just return as-is since most values are plain text
    return value;
}

std::string RDPFileParser::to_json(const RDPFileData& data) {
    auto esc = [](const std::string& s) { return json_utils::escape_string(s); };
    auto b = [](bool v) -> const char* { return v ? "true" : "false"; };

    return std::format(
        R"({{"full_address":"{}","alternate_full_address":"{}","server_port":{},)" 
        R"("username":"{}","domain":"{}","prompt_for_credentials":{},)" 
        R"("prompt_credential_once":{},"authentication_level":{},)" 
        R"("target_is_aad_joined":{},"enable_rds_aad_auth":{},"aad_tenant_id":"{}",)" 
        R"("gateway_hostname":"{}","gateway_usage_method":{},)" 
        R"("gateway_profile_usage_method":{},"gateway_credentials_source":{},)" 
        R"("gateway_brokering_type":{},)" 
        R"("load_balance_info":"{}","wvd_endpoint_pool":"{}",)" 
        R"("arm_path":"{}","workspace_id":"{}","resource_provider":"{}",)" 
        R"("geo":"{}","diagnostic_service_url":"{}","hub_discovery_geo_url":"{}",)" 
        R"("activity_hint":"{}",)" 
        R"("remote_application_program":"{}","remote_desktop_name":"{}",)" 
        R"("remote_application_mode":{},)" 
        R"("desktop_width":{},"desktop_height":{},"screen_mode_id":{},)" 
        R"("smart_sizing":{},"dynamic_resolution":{},"single_mon_in_windowed_mode":{},)" 
        R"("redirect_clipboard":{},"redirect_printers":{},"redirect_smart_cards":{},)" 
        R"("redirect_com_ports":{},"redirect_location":{},)" 
        R"("drives_to_redirect":"{}","devices_to_redirect":"{}",)" 
        R"("cameras_to_redirect":"{}","usb_devices_to_redirect":"{}",)" 
        R"("audio_mode":{},"audio_capture_mode":{},)" 
        R"("is_avd_connection":{},"uses_gateway":{},"requires_aad_auth":{},)" 
        R"("display_name":"{}"}})",
        esc(data.full_address), esc(data.alternate_full_address), data.server_port,
        esc(data.username), esc(data.domain), b(data.prompt_for_credentials),
        b(data.prompt_credential_once), data.authentication_level,
        b(data.target_is_aad_joined), b(data.enable_rds_aad_auth), esc(data.aad_tenant_id),
        esc(data.gateway_hostname), data.gateway_usage_method,
        data.gateway_profile_usage_method, data.gateway_credentials_source,
        data.gateway_brokering_type,
        esc(data.load_balance_info), esc(data.wvd_endpoint_pool),
        esc(data.arm_path), esc(data.workspace_id), esc(data.resource_provider),
        esc(data.geo), esc(data.diagnostic_service_url), esc(data.hub_discovery_geo_url),
        esc(data.activity_hint),
        esc(data.remote_application_program), esc(data.remote_desktop_name),
        data.remote_application_mode,
        data.desktop_width, data.desktop_height, data.screen_mode_id,
        b(data.smart_sizing), b(data.dynamic_resolution), data.single_mon_in_windowed_mode,
        b(data.redirect_clipboard), b(data.redirect_printers), b(data.redirect_smart_cards),
        b(data.redirect_com_ports), b(data.redirect_location),
        esc(data.drives_to_redirect), esc(data.devices_to_redirect),
        esc(data.cameras_to_redirect), esc(data.usb_devices_to_redirect),
        data.audio_mode, data.audio_capture_mode,
        b(data.is_avd_connection()), b(data.uses_gateway()), b(data.requires_aad_auth()),
        esc(data.get_display_name())
    );
}

File: src/rdp_file_parser.rs
//! .rdp file import and parsing.
//!
//! Parses RDP file content (INI-like `key:type:value` format) into
//! [`RdpFileData`] and then maps it to [`ConnectionProfile`].
//! Used for manual RDP file import and WVD feed resource import.
//!
//! Equivalent to: `src/rdp_file_parser.cpp` / `rdp_file_parser.hpp`

use std::path::Path;

use crate::types::ConnectionProfile;

// ---------------------------------------------------------------------------
// Parsed intermediate representation
// ---------------------------------------------------------------------------

/// All supported .rdp / .rdpw file properties, mirroring the C++ `RDPFileData`.
#[derive(Debug, Clone)]
pub struct RdpFileData {
    // Basic connection
    pub full_address: String,
    pub alternate_full_address: String,
    pub server_port: i32,

    // Authentication
    pub username: String,
    pub domain: String,
    pub prompt_for_credentials: bool,
    pub prompt_credential_once: bool,
    pub authentication_level: i32,

    // Azure AD
    pub target_is_aad_joined: bool,
    pub enable_rds_aad_auth: bool,
    pub aad_tenant_id: String,

    // RD Gateway
    pub gateway_hostname: String,
    pub gateway_usage_method: i32,
    pub gateway_profile_usage_method: i32,
    pub gateway_credentials_source: i32,
    pub gateway_brokering_type: i32,

    // AVD / WVD
    pub load_balance_info: String,
    pub wvd_endpoint_pool: String,
    pub arm_path: String,
    pub workspace_id: String,
    pub resource_provider: String,
    pub geo: String,
    pub diagnostic_service_url: String,
    pub hub_discovery_geo_url: String,
    pub activity_hint: String,

    // Remote Application
    pub remote_application_program: String,
    pub remote_desktop_name: String,
    pub remote_application_mode: i32,

    // Display
    pub desktop_width: i32,
    pub desktop_height: i32,
    pub screen_mode_id: i32,
    pub smart_sizing: bool,
    pub dynamic_resolution: bool,
    pub single_mon_in_windowed_mode: i32,

    // Redirection
    pub redirect_clipboard: bool,
    pub redirect_printers: bool,
    pub redirect_smart_cards: bool,
    pub redirect_com_ports: bool,
    pub redirect_location: bool,
    pub drives_to_redirect: String,
    pub devices_to_redirect: String,
    pub cameras_to_redirect: String,
    pub usb_devices_to_redirect: String,

    // Audio
    pub audio_mode: i32,
    pub audio_capture_mode: i32,

    // Security / signature
    pub sign_scope: String,
    pub signature: String,
}

impl Default for RdpFileData {
    fn default() -> Self {
        Self {
            full_address: String::new(),
            alternate_full_address: String::new(),
            server_port: 3389,

            username: String::new(),
            domain: String::new(),
            prompt_for_credentials: true,
            prompt_credential_once: false,
            authentication_level: 2,

            target_is_aad_joined: false,
            enable_rds_aad_auth: false,
            aad_tenant_id: String::new(),

            gateway_hostname: String::new(),
            gateway_usage_method: 0,
            gateway_profile_usage_method: 0,
            gateway_credentials_source: 0,
            gateway_brokering_type: 0,

            load_balance_info: String::new(),
            wvd_endpoint_pool: String::new(),
            arm_path: String::new(),
            workspace_id: String::new(),
            resource_provider: String::new(),
            geo: String::new(),
            diagnostic_service_url: String::new(),
            hub_discovery_geo_url: String::new(),
            activity_hint: String::new(),

            remote_application_program: String::new(),
            remote_desktop_name: String::new(),
            remote_application_mode: 0,

            desktop_width: 0,
            desktop_height: 0,
            screen_mode_id: 1,
            smart_sizing: false,
            dynamic_resolution: false,
            single_mon_in_windowed_mode: 0,

            redirect_clipboard: true,
            redirect_printers: false,
            redirect_smart_cards: false,
            redirect_com_ports: false,
            redirect_location: false,
            drives_to_redirect: String::new(),
            devices_to_redirect: String::new(),
            cameras_to_redirect: String::new(),
            usb_devices_to_redirect: String::new(),

            audio_mode: 0,
            audio_capture_mode: 0,

            sign_scope: String::new(),
            signature: String::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// Helper methods on RdpFileData
// ---------------------------------------------------------------------------

impl RdpFileData {
    /// True when the file describes an Azure Virtual Desktop / WVD session.
    pub fn is_avd_connection(&self) -> bool {
        !self.wvd_endpoint_pool.is_empty()
            || !self.arm_path.is_empty()
            || !self.load_balance_info.is_empty()
            || self.enable_rds_aad_auth
    }

    /// True when an RD Gateway is configured and enabled.
    pub fn uses_gateway(&self) -> bool {
        !self.gateway_hostname.is_empty() && self.gateway_usage_method > 0
    }

    /// True when Azure AD authentication is required.
    pub fn requires_aad_auth(&self) -> bool {
        self.target_is_aad_joined || self.enable_rds_aad_auth
    }

    /// Effective connection host (gateway for AVD, full_address otherwise).
    pub fn get_connection_host(&self) -> &str {
        if self.is_avd_connection() && self.uses_gateway() {
            return &self.gateway_hostname;
        }
        // Strip an embedded `:port` suffix if present.
        self.full_address
            .rsplit_once(':')
            .map_or(self.full_address.as_str(), |(host, _)| host)
    }

    /// Display name for the UI (remote desktop name or hostname).
    pub fn get_display_name(&self) -> &str {
        if !self.remote_desktop_name.is_empty() {
            return &self.remote_desktop_name;
        }
        &self.full_address
    }

    /// Convert the parsed data into a [`ConnectionProfile`].
    pub fn to_connection_profile(&self) -> ConnectionProfile {
        // Separate host and optional embedded port from "full address".
        let (hostname, embedded_port) = split_host_port(&self.full_address);

        let port = if self.server_port != 3389 && self.server_port > 0 {
            self.server_port as u16
        } else if let Some(p) = embedded_port {
            p
        } else {
            3389
        };

        ConnectionProfile {
            name: self.get_display_name().to_owned(),
            hostname: hostname.to_owned(),
            port,
            username: self.username.clone(),
            domain: self.domain.clone(),
            width: if self.desktop_width > 0 {
                self.desktop_width as u32
            } else {
                1920
            },
            height: if self.desktop_height > 0 {
                self.desktop_height as u32
            } else {
                1080
            },
            fullscreen: self.screen_mode_id == 2,
            clipboard: self.redirect_clipboard,
            home_drive: !self.drives_to_redirect.is_empty(),
            dynamic_resolution: self.dynamic_resolution,
            gateway_hostname: self.gateway_hostname.clone(),
            enable_rds_aad_auth: self.enable_rds_aad_auth,
            target_is_aad_joined: self.target_is_aad_joined,
            load_balance_info: self.load_balance_info.clone(),
            aad_tenant_id: self.aad_tenant_id.clone(),
            wvd_endpoint_pool: self.wvd_endpoint_pool.clone(),
            workspace_id: self.workspace_id.clone(),
            arm_path: self.arm_path.clone(),
            remote_application_program: self.remote_application_program.clone(),
            remote_desktop_name: self.remote_desktop_name.clone(),
            ..ConnectionProfile::default()
        }
    }
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

/// Parse the text content of an `.rdp` file into [`RdpFileData`].
///
/// Lines that are empty, start with `#` or `;`, or do not match the
/// `key:type:value` format are silently skipped.
pub fn parse_content(content: &str) -> Result<RdpFileData, String> {
    let mut data = RdpFileData::default();

    for line in content.lines() {
        let line = line.trim_end_matches('\r');
        if line.is_empty() || line.starts_with('#') || line.starts_with(';') {
            continue;
        }

        let (key, _type_ch, value) = match parse_line(line) {
            Some(t) => t,
            None => continue,
        };

        let key_lower = key.to_ascii_lowercase();
        apply_field(&mut data, &key_lower, value);
    }

    if data.full_address.is_empty() && data.gateway_hostname.is_empty() {
        return Err("No valid connection address found".into());
    }

    Ok(data)
}

/// Read an `.rdp` file from disk, strip a UTF-8 BOM if present, and parse.
pub fn parse_file(path: &Path) -> Result<RdpFileData, String> {
    let bytes = std::fs::read(path).map_err(|e| format!("Failed to open file: {e}"))?;

    // Strip UTF-8 BOM (EF BB BF) if present.
    let content = if bytes.starts_with(&[0xEF, 0xBB, 0xBF]) {
        std::str::from_utf8(&bytes[3..])
    } else {
        std::str::from_utf8(&bytes)
    }
    .map_err(|e| format!("Invalid UTF-8: {e}"))?;

    parse_content(content)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/// Split a single RDP line into `(key, type_char, value)`.
fn parse_line(line: &str) -> Option<(&str, char, &str)> {
    let first = line.find(':')?;
    let rest = &line[first + 1..];
    let second = rest.find(':')?;

    let key = &line[..first];
    let type_str = &rest[..second];
    let type_ch = type_str.chars().next()?;
    let value = &rest[second + 1..];

    Some((key, type_ch, value))
}

/// Apply a parsed key/value pair to the correct field on [`RdpFileData`].
fn apply_field(data: &mut RdpFileData, key: &str, value: &str) {
    // Helpers
    let as_bool = || value == "1";
    let as_int = || value.parse::<i32>().unwrap_or(0);

    match key {
        // Basic connection
        "full address" => data.full_address = value.to_owned(),
        "alternate full address" => data.alternate_full_address = value.to_owned(),
        "server port" => data.server_port = as_int(),
        "username" => data.username = value.to_owned(),
        "domain" => data.domain = value.to_owned(),

        // Authentication
        "prompt for credentials" => data.prompt_for_credentials = as_bool(),
        "promptcredentialonce" => data.prompt_credential_once = as_bool(),
        "authentication level" => data.authentication_level = as_int(),

        // Azure AD
        "targetisaadjoined" => data.target_is_aad_joined = as_bool(),
        "enablerdsaadauth" => data.enable_rds_aad_auth = as_bool(),
        "aadtenantid" => data.aad_tenant_id = value.to_owned(),

        // Gateway
        "gatewayhostname" => data.gateway_hostname = value.to_owned(),
        "gatewayusagemethod" => data.gateway_usage_method = as_int(),
        "gatewayprofileusagemethod" => data.gateway_profile_usage_method = as_int(),
        "gatewaycredentialssource" => data.gateway_credentials_source = as_int(),
        "gatewaybrokeringtype" => data.gateway_brokering_type = as_int(),

        // AVD / WVD
        "loadbalanceinfo" => data.load_balance_info = value.to_owned(),
        "wvd endpoint pool" => data.wvd_endpoint_pool = value.to_owned(),
        "armpath" => data.arm_path = value.to_owned(),
        "workspace id" => data.workspace_id = value.to_owned(),
        "resourceprovider" => data.resource_provider = value.to_owned(),
        "geo" => data.geo = value.to_owned(),
        "diagnosticserviceurl" => data.diagnostic_service_url = value.to_owned(),
        "hubdiscoverygeourl" => data.hub_discovery_geo_url = value.to_owned(),
        "activityhint" => data.activity_hint = value.to_owned(),

        // Remote Application
        "remoteapplicationprogram" => data.remote_application_program = value.to_owned(),
        "remotedesktopname" => data.remote_desktop_name = value.to_owned(),
        "remoteapplicationmode" => data.remote_application_mode = as_int(),

        // Display
        "desktopwidth" => data.desktop_width = as_int(),
        "desktopheight" => data.desktop_height = as_int(),
        "screen mode id" => data.screen_mode_id = as_int(),
        "smart sizing" => data.smart_sizing = as_bool(),
        "dynamic resolution" => data.dynamic_resolution = as_bool(),
        "singlemoninwindowedmode" => data.single_mon_in_windowed_mode = as_int(),

        // Redirection
        "redirectclipboard" => data.redirect_clipboard = as_bool(),
        "redirectprinters" => data.redirect_printers = as_bool(),
        "redirectsmartcards" => data.redirect_smart_cards = as_bool(),
        "redirectcomports" => data.redirect_com_ports = as_bool(),
        "redirectlocation" => data.redirect_location = as_bool(),
        "drivestoredirect" => data.drives_to_redirect = value.to_owned(),
        "devicestoredirect" => data.devices_to_redirect = value.to_owned(),
        "camerastoredirect" => data.cameras_to_redirect = value.to_owned(),
        "usbdevicestoredirect" => data.usb_devices_to_redirect = value.to_owned(),

        // Audio
        "audiomode" => data.audio_mode = as_int(),
        "audiocapturemode" => data.audio_capture_mode = as_int(),

        // Security
        "signscope" => data.sign_scope = value.to_owned(),
        "signature" => data.signature = value.to_owned(),

        _ => {} // Unknown keys are silently ignored.
    }
}

/// Split `"host:port"` into `(host, Some(port))` or `("host", None)`.
fn split_host_port(address: &str) -> (&str, Option<u16>) {
    if let Some((host, port_str)) = address.rsplit_once(':') {
        if let Ok(port) = port_str.parse::<u16>() {
            return (host, Some(port));
        }
    }
    (address, None)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_basic_rdp() {
        let content = "\
full address:s:server.example.com
server port:i:3390
username:s:user@domain.com
domain:s:CORP
desktopwidth:i:1920
desktopheight:i:1080
screen mode id:i:2
redirectclipboard:i:1
";
        let data = parse_content(content).unwrap();
        assert_eq!(data.full_address, "server.example.com");
        assert_eq!(data.server_port, 3390);
        assert_eq!(data.username, "user@domain.com");
        assert_eq!(data.domain, "CORP");
        assert_eq!(data.desktop_width, 1920);
        assert_eq!(data.desktop_height, 1080);
        assert_eq!(data.screen_mode_id, 2);
        assert!(data.redirect_clipboard);
    }

    #[test]
    fn parse_avd_rdp() {
        let content = "\
full address:s:rdgateway.wvd.microsoft.com
gatewayhostname:s:rdgateway.wvd.microsoft.com
gatewayusagemethod:i:1
loadbalanceinfo:s:tsv://MS Terminal Services Plugin.1.wvd
targetisaadjoined:i:1
enablerdsaadauth:i:1
workspace id:s:ws-123
";
        let data = parse_content(content).unwrap();
        assert!(data.is_avd_connection());
        assert!(data.uses_gateway());
        assert!(data.requires_aad_auth());
        assert_eq!(data.get_connection_host(), "rdgateway.wvd.microsoft.com");
    }

    #[test]
    fn skips_comments_and_empty_lines() {
        let content = "\
# comment
; another comment

full address:s:host.test
";
        let data = parse_content(content).unwrap();
        assert_eq!(data.full_address, "host.test");
    }

    #[test]
    fn rejects_missing_address() {
        let content = "username:s:foo\n";
        assert!(parse_content(content).is_err());
    }

    #[test]
    fn to_connection_profile_basic() {
        let content = "\
full address:s:myhost.com:3390
screen mode id:i:2
drivestoredirect:s:*
redirectclipboard:i:1
";
        let data = parse_content(content).unwrap();
        let prof = data.to_connection_profile();
        assert_eq!(prof.hostname, "myhost.com");
        assert_eq!(prof.port, 3390);
        assert!(prof.fullscreen);
        assert!(prof.home_drive);
        assert!(prof.clipboard);
    }

    #[test]
    fn display_name_prefers_remote_desktop_name() {
        let content = "\
full address:s:host.test
remotedesktopname:s:My Remote Desktop
";
        let data = parse_content(content).unwrap();
        assert_eq!(data.get_display_name(), "My Remote Desktop");
    }

    #[test]
    fn display_name_falls_back_to_address() {
        let content = "full address:s:fallback.host\n";
        let data = parse_content(content).unwrap();
        assert_eq!(data.get_display_name(), "fallback.host");
    }

    #[test]
    fn gateway_only_file_accepted() {
        let content = "\
gatewayhostname:s:gw.example.com
gatewayusagemethod:i:1
";
        let data = parse_content(content).unwrap();
        assert!(data.uses_gateway());
    }

    #[test]
    fn embedded_port_in_full_address() {
        let content = "full address:s:server.com:5555\n";
        let data = parse_content(content).unwrap();
        let prof = data.to_connection_profile();
        assert_eq!(prof.hostname, "server.com");
        assert_eq!(prof.port, 5555);
    }

    #[test]
    fn server_port_overrides_embedded() {
        let content = "\
full address:s:server.com:5555
server port:i:6666
";
        let data = parse_content(content).unwrap();
        let prof = data.to_connection_profile();
        assert_eq!(prof.port, 6666);
    }
}
